OSDOCS-3898: vSphere Topology Awareness (#52116)
openshift/openshift-docs: modules/persistent-storage-csi-vsphere-top-aware.adoc

// Module included in the following assemblies:
//
// persistent-storage-csi-vsphere.adoc
//

:content-type: PROCEDURE
[id="persistent-storage-csi-vsphere-top-aware_{context}"]
= vSphere Topology Awareness

{product-title} provides the ability to deploy {product-title} for vSphere on different zones and regions, which allows you to deploy over multiple data centers and clusters, thus helping to avoid a single-point-of-failure.

.Procedure
Use the following procedure to deploy {product-title} for vSphere storage on different zones and regions.

. In the VMware vCenter vSphere client GUI:
.. Create and assign `openshift-zone` and `openshift-region` tags as needed to the hosts on the desired clusters.

> .. Create and apply `cat-zone-local-datastore` tags to the desired datastores.

Author: Makes sense. Done.

+
For more information about vSphere tags, see the WMware vSphere documentation.

. In the `clustercsidriver` object YAML file, edit the `driverConfig` section:
* Specify the `openshift-zone` and `openshift-region` tags that you created earlier.
* Set `driverType` to `vSphere`.
+
[source, terminal]
----
~ $ oc edit clustercsidriver csi.vsphere.vmware.com -o yaml
----
+
Example output:
+
[source, terminal]
----
apiVersion: operator.openshift.io/v1
kind: ClusterCSIDriver
metadata:
 name: csi.vsphere.vmware.com
spec:
 logLevel: Normal
 managementState: Managed
 observedConfig: null
 operatorLogLevel: Normal
 unsupportedConfigOverrides: null
  driverConfig: 
    driverType: vSphere <1>
    vSphere:
      topologyCategories: <2>
        - openshift-zone

>         - openshift-region

Author: Done.

----
<1> Ensure that `driverType` is set to `vSphere`.

> <2> `openshift-zone` and `openshift-region` tags created earlier in vCenter.

Author: @duanwei33 @gnufied So you are looking for a global change with this? **tags** -> **categories**?

> <2> `openshift-zone` and `openshift-region` tags created earlier in vCenter.

Reviewer: @lpettyjo , actually categories: tags is more like a key: value pair, `openshift-zone` and `openshift-region` are categories and we could specify like `zone-a` and `region-a` as assigned tags. They are defined in vSphere side and used in different places in CSI Driver. This is why I have another commit https://github.com/openshift/openshift-docs/pull/52116#discussion_r1040389078.
Please refer to first two steps of `vSphere CSI driver topology doc` shared by Hemant.

> <2> `openshift-zone` and `openshift-region` tags created earlier in vCenter.

Reviewer 2: Yes i agree with @duanwei33

> <2> `openshift-zone` and `openshift-region` tags created earlier in vCenter.

Author: Done.

. Verify that `CSINode` object has topology keys:
+
[source, terminal]
----
~ $ oc get csinode
----
+
Example output:
+
[source, terminal]
----
NAME                        DRIVERS     AGE
co8-4s88d-infra-2m5vd       1           27m
co8-4s88d-master-0          1           70m
co8-4s88d-master-1          1           70m
co8-4s88d-master-2          1           70m
co8-4s88d-worker-j2hmg      1           47m
co8-4s88d-worker-mbb46      1           47m
co8-4s88d-worker-zlk7d      1           47m
----
+
[source, terminal]
----
~ $ oc get csinode co8-4s88d-worker-j2hmg -o yaml
----
+
Exmaple output:
+
[source, terminal]
----
...
spec:
    drivers:
    - allocatable:
        count: 59
    name: csi-vsphere.vmware.com
    nodeID: co8-4s88d-worker-j2hmg
    topologyKeys: <1>
    - topology.csi-vmware.com/openshift-zone

>     - topology.csi-vmware.com/openshift-region

Author: Done.

----
<1> Topology keys from vSphere `openshift-zone` and `openshift-region` catagory tags.
+
[NOTE]
=====
`CSINode` objects might take some time to receive updated topology information. After the driver is updated, `CSINode` objects should have topology keys in them.
=====

. Create a storage policy that uses the category tags:
.. In vCenter, from the main menu, click *Policies and Profiles*.
.. On the *Policies and Profiles* page, in the left navigation pane, click *VM Storage Policies*.
.. Click *CREATE*.
.. Type a name for the storage policy.

> .. For the rules, select tags that target the desired datastore.

Author: Done. PTAL.

. Create a new storage class that uses the new zoned storage policy:
.. Click *Storage* > *StorageClasses*.
.. On the *StorageClasses* page, click *Create StorageClass*.
.. Type a name for the new storage class in *Name*.
.. Under *Provisioner*, select *csi.vsphere.vmware.com*. 
.. Under *Additional parameters*, for the StoragePolicyName parameter, set *Value* to the name of the new zoned storage policy that you created earlier.
.. Click *Create*.
+

> For example:

Reviewer: I think we could mention it could be created by CLI (And we have given the example yaml file)

Author: Done.

+
[source, yaml]
----
allowVolumeExpansion: true
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
 name: zoned-sc <1>
parameters:
 StoragePolicyName:  zoned-storage-policy <2>
provisioner: csi.vsphere.vmware.com
reclaimPolicy: Delete
volumeBindingMode: WaitForFirstConsumer
----
<1> New topology aware storage class name.
<2> Specify zoned storage policy.

.Results
Creating persistent volume claims (PVCs) and persistent volumes (PVs) from the topology aware storage class are truly zonal, and should use the datastore in their respective zone depending on how pods are scheduled.

For example:

[source, terminal]
----
~ $ oc get pv <pv-name> -o yaml
----

[source, terminal]
----
...
nodeAffinity:
  required:
    nodeSelectorTerms:
    - matchExpressions:
      - key: topology.csi.vmware.com/openshift-zone <1>
        operator: In  
        values:

>         - openshift-zone

Reviewer: ```suggestion
        - <openshift-zone>
```

This should be the specified tag name, I think <openshift-zone> or <openshift-zone name> should be more clear.

Author: I think that we are sticking with those tags. See this comment https://github.com/openshift/openshift-docs/pull/52116#discussion_r1021886744.

Reviewer: @lpettyjo Here we need to give a specific tag name for openshift-zone like zone-a/zone-b, so I suggest to use <openshift-zone> or <openshift-zone-name>, cc @gnufied

Author: Done.

      -key: topology.csi.vmware.com/openshift-region <1>
        operator: In
        values:

>         - openshift-region

Reviewer: ```suggestion
        - <openshift-region>
```
This should be the specified tag name, I think <openshift-region> or <openshift-region name> should be more clear.

Author: I think that we are sticking with those tags. See Hemant's comment https://github.com/openshift/openshift-docs/pull/52116#discussion_r1021886744.

Author: Done.

Author: Done.

// Module included in the following assemblies:
//
// persistent-storage-csi-vsphere.adoc
//

:content-type: PROCEDURE
[id="persistent-storage-csi-vsphere-top-aware_{context}"]
= vSphere Topology Awareness

{product-title} provides the ability to deploy {product-title} for vSphere on different zones and regions, which allows you to deploy over multiple data centers and clusters, thus helping to avoid a single-point-of-failure.

:FeatureName: vSphere topology Awareness

> include::snippets/technology-preview.adoc[leveloffset=+1]

Reviewer 2: It's not TP

Author: Fixed in doc and RN.

.Procedure

> To deploy {product-title} for vSphere storage on different zones and regions:

Reviewer 3: Regarding the colon at the end here: Use a colon after the introduction to a vertical list, including a procedure or substeps in a procedure. To avoid translation problems, make the introduction a complete sentence. Note that I only think this applies to the introduction sentence, not the following (i.e., 1., 2., etc.). 

See: https://www.ibm.com/docs/en/ibm-style?topic=punctuation-colons

Could we reword this as something like: 

Use the following procedure to deploy (product-title} for vSphere storage on different zones and regions:

. In the VMware vCenter vSphere client GUI:

> .. Create and assign zone and region type tags as needed to the hosts on the desired clusters.

Reviewer 3: In the docs, we often put tag types in backticks. For example, `redis` tags, `latest` tag, etc. I'm not sure if you want to do that here. Just calling it out.

> .. Create and apply "cat-zone-local-datastore" tags to the desired datastores.

Reviewer 3: Quotes (") or backticks (`)  for `cat-zone-local-datastore` tags? If we put this one in backticks, we might want to put `region` and `zone` (step A above) in backticks.

> .. Create and apply `cat-zone-local-datastore` tags to the desired datastores.

Author: Let me know how you want to proceed here.

+

> For more information about vSphere tags, see the WMware vSphere documentation.

Reviewer 3: Should we link to the VMWare vSphere documentation? link:URL[WMware vSphere documentation]? 

Edit: I see it's included below.

. In the `clustercsidriver` object YAML file, edit the "driverConfig" section:

> * Specify the zone and region tags that you created earlier.

Reviewer 3: Here, after Step 2, we're using substeps as bullet points. Above, we're using substeps as letters. Is there a reason? Maybe users selection option a) or b) above, or something?

> * Set "driverType" to "vSphere".

Reviewer 3: Quotes (") or backticks (`) for driverType and vSphere?

+
[source, terminal]
----

> ~ $ oc get clustercsidriver csi.vsphere.vmware.com -o yaml

Reviewer 3: I'm sure why/when we started putting ~ before the $ in commands?

Author: That's the command prompt.

> ~ $ oc get clustercsidriver csi.vsphere.vmware.com -o yaml

Reviewer 2: oc get won't edit the object

Author: Done.

----
+
Example output:
+
[source, terminal]
----
apiVersion: operator.openshift.io/v1
kind: ClusterCSIDriver
metadata:
 name: csi.vsphere.vmware.com
spec:
 logLevel: Normal
 managementState: Managed
 observedConfig: null
 operatorLogLevel: Normal
 unsupportedConfigOverrides: null
  driverConfig: 
    driverType: vSphere <1>
    vSphere:

>       topologyCategories: <2>

Reviewer 2: @gnufied do we want to restrict the tags naming?

Reviewer 4: Yeah - lets stick to `openshift-zone` and `openshift-region` as topology category names in our docs.

Author: Done.

Author: Done.

        - k8s-zone
        - k8s-region
----

> <1> Ensure that "driverType" is set to "vSphere".

Reviewer 3: Same here, re: driverType and vSphere

> <2> Zone and region tags created earlier in vCenter.

Reviewer 3: If choosing to put Zone/region in tags, it'll need updated here.

> . Verify that `CSINode` object has topology keys:

Reviewer 3: the `CSINode` object? other examples in the docs say "the CSINode object" (without CSINode being in backticks). Might be situational.

Author: I checked and the file has every instance of "csinode" with backticks. They must just not be rendering as expected in the preview.

+
[source, terminal]
----
~ $ oc get csinode
----
+
Example output:
+
[source, terminal]
----
NAME                        DRIVERS     AGE
co8-4s88d-infra-2m5vd       1           27m
co8-4s88d-master-0          1           70m
co8-4s88d-master-1          1           70m
co8-4s88d-master-2          1           70m
co8-4s88d-worker-j2hmg      1           47m
co8-4s88d-worker-mbb46      1           47m
co8-4s88d-worker-zlk7d      1           47m
----
+
[source, terminal]
----
~ $ oc get csinode co8-4s88d-worker-j2hmg -o yaml
----
+
Exmaple output:
+
[source, terminal]
----
...
spec:
    drivers:
    - allocatable:
        count: 59
    name: csi-vsphere.vmware.com
    nodeID: co8-4s88d-worker-j2hmg
    topologyKeys: <1>
    - topology.csi-vmware.com/k8s-zone
    - topology.csi-vmware.com/k8s-region

> ----

Reviewer 4: We will have to update this to reflect `openshift-zone` and `openshift-region`

Author: Done

<1> Topology keys from vSphere zone and region catagory tags.
+
[NOTE]
=====

> `CSINode` objects may take some time to receive updated topology information. After the driver is updated, `CSINode` objects should have topology keys in them.

Reviewer 3: might take some time 

(IMBSG recommends against using "may")

=====

. Create a storage policy that uses the category tags:
.. In vCenter, from the main menu, click *Policies and Profiles*.

> .. On the *Policies and Profiles* page, in the left pane, click *VM Storage Policies*.

Reviewer 3: IBMSG recommends against using directionals (i.e, "left pane."). Maybe just "navigation pane" ?

.. Click *CREATE*.
.. Type a name for the storage policy.
.. For the rules, select tags that target the desired datastore.

. Create a new storage class that uses the new zoned storage policy:

> .. Click *Storage* > *StorageClasses*.

Reviewer 3: Here, we go back to letters (above we did bullets and letters). Maybe it was intentional?

Author: letters for sublevel ordered steps and bullets for unordered steps.

.. On the *StorageClasses* page, click *Create StorageClass*.
.. Type a name for the new storage class in *Name*.
.. Under *Provisioner*, select *csi.vsphere.vmware.com*. 

> .. Under *Additional parameters*, for the StoragePolicyName parameter, set *Value* to the name of the new zoned storage policy that you created earlier.

Reviewer 3: Should StoragePolicyName be in backticks?

// Module included in the following assemblies:
//
// persistent-storage-csi-vsphere.adoc
//

:content-type: PROCEDURE
[id="persistent-storage-csi-vsphere-top-aware_{context}"]
= vSphere Topology Awareness

> {product-title} provides the ability to deploy {product-title} for vSphere on different zones and regions, which allows you to deploy over multiple data centers and clusters, thus helping to avoid a single-point-of-failure.

Reviewer 5: per discussion I thought we weren't supporting multiple vcenter datacenters

Author: @gnufied ?

Reviewer 4: Yeah we should update the wording. OCP <=4.12 does not support multiple datacenters. We should revisit this doc for 4.13 though, where we do support multiple datacenters properly.

Author: Fixed. Thanks! https://github.com/openshift/openshift-docs/pull/52116#issuecomment-1375713882

.Procedure
Use the following procedure to deploy {product-title} for vSphere storage on different zones and regions.

. In the VMware vCenter vSphere client GUI:

> .. Create and assign `openshift-zone` and `openshift-region` tags as needed to the hosts on the desired clusters.

Reviewer 2: openshift-zone & openshift-regions are categories not tags.

Author: Done.

> .. Create and assign `openshift-zone` and `openshift-region` category tags as needed to the hosts on the desired clusters.

Reviewer 5: Switch `category tags` to `tag categories`
link vmware's docs to create tags and tag categories 

https://docs.vmware.com/en/VMware-vSphere/7.0/com.vmware.vsphere.vcenterhost.doc/GUID-E8E854DD-AA97-4E0C-8419-CE84F93C4058.html

Reviewer 5: This section is confusing to me.
You don't assign tag categories to vcenter objects. You create tags that are associated with a category. The tag is attached to either a vcenter datacenter or cluster.

See the zonal presentation for 4.12 TP:
https://docs.google.com/presentation/d/1AAuGLYRQ93rCS-6nKrcHaBTNqt9wNvwkfNWR66n3i90/edit#slide=id.g14cc0ab5502_0_44

Author: Fixed.

Author: There is a link to vwmare's doc for tags/categories in the Additional Resources section.

> .. Create and assign `cat-zone-local-datastore` tags to the desired datastores.

Reviewer 2: Same here, show an example

Author: See comment above.

Author: Done.

> .. Create and assign `cat-zone-local-datastore` category tags to the desired datastores.

Reviewer 5: same as above should be `tag category` 

And what is this tag and tag category for? Not described

Author: Fixed.

+

> For more information about vSphere tags, see the WMware vSphere documentation.

Reviewer 2: I think we should add a small example table for a 6 nodes clusters, 3 masters / 3 worker spread on 3 zones. Show what category / tag the customer should use.

Author: Done.

> For more information about vSphere tags, see the WMware vSphere documentation.

Reviewer: ```suggestion
For more information about vSphere tags, see the VMware vSphere documentation.
```

Author: Fixed.

----
<1> Topology keys from vSphere `openshift-zone` and `openshift-region` catagory tags.
+
[NOTE]
=====
`CSINode` objects might take some time to receive updated topology information. After the driver is updated, `CSINode` objects should have topology keys in them.
=====

. Create a storage policy that uses the category tags:
.. In vCenter, from the main menu, click *Policies and Profiles*.
.. On the *Policies and Profiles* page, in the left navigation pane, click *VM Storage Policies*.
.. Click *CREATE*.
.. Type a name for the storage policy.
.. For the rules, select tags that target the desired datastore. 
+ 
The datastores are listed in the storage compatibility table.

. Create a new storage class that uses the new zoned storage policy:
.. Click *Storage* > *StorageClasses*.
.. On the *StorageClasses* page, click *Create StorageClass*.
.. Type a name for the new storage class in *Name*.
.. Under *Provisioner*, select *csi.vsphere.vmware.com*. 
.. Under *Additional parameters*, for the StoragePolicyName parameter, set *Value* to the name of the new zoned storage policy that you created earlier.
.. Click *Create*.
+
For example:
+
[source, yaml]
----

> allowVolumeExpansion: true

Reviewer 2: For clarity, we usually start the storage class definition with 
kind: StorageClass

Any reason to start with something else?

Author: Fixed.

[id="persistent-storage-csi-vsphere-top-aware_{context}"]
= vSphere Topology Awareness

{product-title} provides the ability to deploy {product-title} for vSphere on different zones and regions, which allows you to deploy over multiple data centers and clusters, thus helping to avoid a single point of failure. 

This is accomplished by defining zone and region categories in vCenter, and then assigning these categories to different failure domains, such as a compute cluster, host, or data center, by creating tags for these zone and region categories. After you have created the appropriate categories, and assigned tags to vCenter objects, you can create additional machinesets that create virtual machines (VMs) that are responsible for scheduling pods in those failure domains.

.Procedure
. In the VMware vCenter vSphere client GUI, define appropriate zone and region catagories and tags.
+
Zones and regions in the vSphere environment are typically defined by creating categories called `openshift-zone` and `openshift-region`.
Usually, category names can be any valid string, but {product-title} recommends use of `openshift-zone` and `openshift-region`
categories. 

> +

Reviewer 4: Lets reword this to be more stronger in meaning - "while vSphere allows creation of categories with any arbitrary name, {product-title} strongly recommends use of `openshift-region` and `openshift-zone` names for defining topology`

Author: Done.

> The following example defines two failure domains with one region and two zones:

Reviewer 3: Should we say "The following table" since it's a table?

Author: I want to emphasis that this is example data--you can set this up many possible different ways, so I'm going to leave it.

+
.vSphere topology with one region and two zones
|===
|Compute cluster | Failure domain |Description

|Compute cluster: ocp1, 

> Data center: Atlanta

Reviewer 4: In terms of vCenter - I think `Datacenter` is one word.

Reviewer 4: I meant `Datacenter` as vCenter object. It is not same as generic "Data center" word that gets used in the enterprise world. We should stick to termininology inside vcenter. 
<img width="652" alt="Screenshot 2023-01-09 at 9 28 25 AM" src="https://user-images.githubusercontent.com/278/211332301-24cb317d-9e00-437f-8014-3815fe24db79.png">

Author: Got it. Fixed.

|openshift-region: us-east-1 (tag), openshift-zone: us-east-1a (tag)
|This defines a failure domain in region us-east-1 with zone us-east-1a.

|Computer cluster: ocp2, 
Data center: Atlanta
|openshift-region: us-east-1 (tag), openshift-zone: us-east-1b (tag)
|This defines a different failure domain within the same region called us-east-1b.
|===
+
For more information about vSphere categories and tags, see the VMware vSphere documentation.

. To allow the container storage interface (CSI) driver to detect this topology, edit the `clusterCSIDriver` object YAML file `driverConfig` section:
* Specify the `openshift-zone` and `openshift-region` categories that you created earlier.

> * Set `driverType` to `vSphere`.

Reviewer 3: Not 100% sure how second level commands should be formatted, but should we use a colon here instead of a period, since what follows is a command? Additionally, should we add "by running the following command:", as per the OCP guidelines? 

https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output

Author: Not sure about this. The colon is for introducing the two things that you need to do after running the command. I'm going to leave it.

+
[source, terminal]
----
~ $ oc edit clustercsidriver csi.vsphere.vmware.com -o yaml
----
+

> Example output:

Reviewer 3: This was brought to my attention a few weeks ago, and I don't think we're super consistent with it, but apparently "Example output" needs to be written as follows: 

.Example output 

See https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output

+
[source, terminal]
----
apiVersion: operator.openshift.io/v1
kind: ClusterCSIDriver
metadata:
 name: csi.vsphere.vmware.com
spec:
 logLevel: Normal
 managementState: Managed
 observedConfig: null
 operatorLogLevel: Normal
 unsupportedConfigOverrides: null
 driverConfig: 
  driverType: vSphere <1>
    vSphere:
      topologyCategories: <2>
      - openshift-zone
      - openshift-region
----
<1> Ensure that `driverType` is set to `vSphere`.
<2> `openshift-zone` and `openshift-region` categories created earlier in vCenter.

> . Verify that `CSINode` object has topology keys:

Reviewer 3: If you decide to implement "run/enter the following command" above, it will need to be consistent and used here. Additionally, since it looks like this step has two commands, you might write it as ". Verify that `CSINode` object has topology keys by entering the following commands" (plural)

+
[source, terminal]
----
~ $ oc get csinode
----
+

> Example output:

Reviewer 3: See .Example output comment above

+
[source, terminal]
----
NAME                        DRIVERS     AGE
co8-4s88d-infra-2m5vd       1           27m
co8-4s88d-master-0          1           70m
co8-4s88d-master-1          1           70m
co8-4s88d-master-2          1           70m
co8-4s88d-worker-j2hmg      1           47m
co8-4s88d-worker-mbb46      1           47m
co8-4s88d-worker-zlk7d      1           47m
----
+
[source, terminal]
----
~ $ oc get csinode co8-4s88d-worker-j2hmg -o yaml
----
+

> Exmaple output:

Reviewer 3: See .Example output comment above. Also, "Example" is misspelled :D

+
[source, terminal]
----
...
spec:
    drivers:
    - allocatable:
        count: 59
    name: csi-vsphere.vmware.com
    nodeID: co8-4s88d-worker-j2hmg
    topologyKeys: <1>
    - topology.csi.vmware.com/openshift-zone
    - topology.csi.vmware.com/openshift-region
----
<1> Topology keys from vSphere `openshift-zone` and `openshift-region` catagories.
+
[NOTE]
=====
`CSINode` objects might take some time to receive updated topology information. After the driver is updated, `CSINode` objects should have topology keys in them.
=====

> . Create a tag to assign to datastores across failure domains:

Reviewer 3: This might not be the case, but I think that the punctuation here should be a period since it doesn't lead to any commands or directly into the next step. Comparatively, Step 5 (the following) goes directly into the substeps. I don't have a strong preference either way.

+
When an {product-title} spans more than one failure domain, the datastore might not be shared across those failure domains, which is where topology-aware provisioning of persistent volumes (PVs) is useful. 
+
.. In vCenter, create a category called `openshift-zonal-datastore-cat`. This example uses the name "openshift-zonal-datastore-cat" for this category; however, you can use any name you like.
.. In vCenter, create a tag called `openshift-zonal-datastore` for the `openshift-zonal-datastore-cat` category.
.. Assign the `openshift-zonal-datastore` tag to each datastore in a failure domain that would be considered for dynamic provisioning.
+
====
NOTE: You can use any names you like for categories and tags. The names used in this example are provided as recommendations. Ensure that the tags and categories that you define uniquely identify only datastores that are shared with all hosts in the {product-title} cluster.
====

. Create a storage policy that targets the tag-based datastores in each failure domain:
.. In vCenter, from the main menu, click *Policies and Profiles*.

> .. On the *Policies and Profiles* page, in the left navigation pane, click *VM Storage Policies*.

Reviewer 3: In an attempt to adhere to https://www.ibm.com/docs/en/ibm-style?topic=medium-global-audiences, I think we should remove "left" and just say "navigation pane." Depending on the screen, it could be up, down, right, etc. Also some people might use screen readers which further complicates things. Just a suggestion

.. Click *CREATE*.
.. Type a name for the storage policy.
.. For the rules, select the tag that targets the desired datastores (in this example, the `openshift-zonal-datastore` tag).
+ 
The datastores are listed in the storage compatibility table.

. Create a new storage class that uses the new zoned storage policy:
.. Click *Storage* > *StorageClasses*.
.. On the *StorageClasses* page, click *Create StorageClass*.
.. Type a name for the new storage class in *Name*.
.. Under *Provisioner*, select *csi.vsphere.vmware.com*. 
.. Under *Additional parameters*, for the StoragePolicyName parameter, set *Value* to the name of the new zoned storage policy that you created earlier.
.. Click *Create*.
+

> For example:

Reviewer 3: I don't recall another time seeing "For example" used at the end of the procedure like this (not to say there isn't! I just probably haven't seen in), but since we're supposed to use ".Example output", I wonder how ".Example" might look here? Just a thought/consideration. No preference

+
[source, yaml]
----
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: zoned-sc <1>
provisioner: csi.vsphere.vmware.com
parameters:
  StoragePolicyName: zoned-storage-policy <2>
reclaimPolicy: Delete
allowVolumeExpansion: true
volumeBindingMode: WaitForFirstConsumer
----
<1> New topology aware storage class name.
<2> Specify zoned storage policy.
+
====
NOTE: You can also create the storage class by editing the preceding YAML file and running the command `oc create -f $FILE`.
====

.Results
Creating persistent volume claims (PVCs) and PVs from the topology aware storage class are truly zonal, and should use the datastore in their respective zone depending on how pods are scheduled.

For example:

[source, terminal]
----
~ $ oc get pv <pv-name> -o yaml
----

> 

Reviewer 3: Is this the output of the above command? If so, it's missing ".Example output"

// Module included in the following assemblies:
//
// persistent-storage-csi-vsphere.adoc
//

:content-type: PROCEDURE
[id="persistent-storage-csi-vsphere-top-aware_{context}"]

> = vSphere Topology Awareness

Reviewer 4: Lets rename this heading to - `Configure vSphere CSI Topology`

Author: Done.

> {product-title} provides the ability to deploy {product-title} for vSphere on different zones and regions, which allows you to deploy over multiple clusters, thus helping to avoid a single point of failure. 

Reviewer 4: rename `multiple clusters` to `multiple compute clusters`.

Author: Done.

[NOTE]
====

> {product-title} vSphere topology awareness does not support multiple vCenter datacenters.

Reviewer 4: err -  this reads bit awkwardly. We should just say:

```
{product-tile} on vSphere does not yet support multiple Datacenters.
```

Author: Done.

====

> This is accomplished by defining zone and region categories in vCenter, and then assigning these categories to different failure domains, such as a compute cluster or host, by creating tags for these zone and region categories. After you have created the appropriate categories, and assigned tags to vCenter objects, you can create additional machinesets that create virtual machines (VMs) that are responsible for scheduling pods in those failure domains.

Reviewer 4: I would drop `or host` from this line.

Author: Done.

> This is accomplished by defining zone and region categories in vCenter, and then assigning these categories to different failure domains, such as a compute cluster, host, or datacenter, by creating tags for these zone and region categories. After you have created the appropriate categories, and assigned tags to vCenter objects, you can create additional machinesets that create virtual machines (VMs) that are responsible for scheduling pods in those failure domains.

Reviewer 4: nit: We should drop datacenter from here too, since we don't support individual datacenter as a failure-domain in <=4.12

.Procedure
. In the VMware vCenter vSphere client GUI, define appropriate zone and region catagories and tags.
+

> While vSphere allows you to create categories with any arbitrary name, {product-title} strongly recommends use of `openshift-region`` and `openshift-zone`` names for defining topology.

Reviewer 4: Here strings like `openshfit-region` etc seems double escaped with backticks. They should be only once escaped.

Author: Done.

+
The following example defines two failure domains with one region and two zones:
+
.vSphere topology with one region and two zones
|===
|Compute cluster | Failure domain |Description

|Compute cluster: ocp1, 
Datacenter: Atlanta
|openshift-region: us-east-1 (tag), openshift-zone: us-east-1a (tag)
|This defines a failure domain in region us-east-1 with zone us-east-1a.

|Computer cluster: ocp2, 
Datacenter: Atlanta
|openshift-region: us-east-1 (tag), openshift-zone: us-east-1b (tag)

> |This defines a different failure domain within the same region called us-east-1b.

Reviewer 2: This defines an us-east-1b zone the same region.

co8-4s88d-master-0          1           70m
co8-4s88d-master-1          1           70m
co8-4s88d-master-2          1           70m
co8-4s88d-worker-j2hmg      1           47m
co8-4s88d-worker-mbb46      1           47m
co8-4s88d-worker-zlk7d      1           47m
----
+
[source, terminal]
----
~ $ oc get csinode co8-4s88d-worker-j2hmg -o yaml
----
+
.Example output
+
[source, terminal]
----
...
spec:
    drivers:
    - allocatable:
        count: 59
    name: csi-vsphere.vmware.com
    nodeID: co8-4s88d-worker-j2hmg
    topologyKeys: <1>
    - topology.csi.vmware.com/openshift-zone
    - topology.csi.vmware.com/openshift-region
----
<1> Topology keys from vSphere `openshift-zone` and `openshift-region` catagories.
+
[NOTE]
=====
`CSINode` objects might take some time to receive updated topology information. After the driver is updated, `CSINode` objects should have topology keys in them.
=====

. Create a tag to assign to datastores across failure domains:
+
When an {product-title} spans more than one failure domain, the datastore might not be shared across those failure domains, which is where topology-aware provisioning of persistent volumes (PVs) is useful. 
+

> .. In vCenter, create a category called `openshift-zonal-datastore-cat`. This example uses the name "openshift-zonal-datastore-cat" for this category; however, you can use any name you like.

Reviewer 4: ```suggestion
.. In vCenter, create a category for tagging the datastores. For example - `openshift-zonal-datastore-cat`.You can use any other category name, as long as category uniquely is used for tagging datastores participating in {product-title} cluster. Also, please make sure that `StoragePod`, `Datastore` and `Folder` are selected as Associable Entities for created category. 
```

Author: Done.

> .. In vCenter, create a tag called `openshift-zonal-datastore` for the `openshift-zonal-datastore-cat` category.

Reviewer 4: ```suggestion
.. In vCenter, create a tag that uses aforementioned category. In this example, we will call it `openshift-zonal-datastore`.

Author: Done.

> .. Assign the `openshift-zonal-datastore` tag to each datastore in a failure domain that would be considered for dynamic provisioning.

Author: Done.

+
[NOTE]
====
You can use any names you like for categories and tags. The names used in this example are provided as recommendations. Ensure that the tags and categories that you define uniquely identify only datastores that are shared with all hosts in the {product-title} cluster.
====

. Create a storage policy that targets the tag-based datastores in each failure domain:
.. In vCenter, from the main menu, click *Policies and Profiles*.
.. On the *Policies and Profiles* page, in the navigation pane, click *VM Storage Policies*.
.. Click *CREATE*.
.. Type a name for the storage policy.

> .. For the rules, select the tag that targets the desired datastores (in this example, the `openshift-zonal-datastore` tag).

Reviewer 4: ```suggestion
.. For the rules, choose Tag Placement rules and select the tag and category that targets the desired datastores (in this example, the `openshift-zonal-datastore` tag).
```

Author: Done.
